src/v/redpanda/admin/server.h

```diff
@@ -258,6 +268,31 @@ class admin_server {
           });
     }
 
+    using future_json_status_function
+      = std::function<ss::future<json_status_return_type>(
+        std::unique_ptr<ss::http::request> req)>;
+
+    template<auth_level required_auth, bool peek_auth = false>
+    void register_status_route(
+      const ss::httpd::path_description& path,
+      future_json_status_function handler) {
+        return register_route_raw_async<required_auth, peek_auth>(
+          path, [handler = std::move(handler)](auto req, auto rep) {
+              return handler(std::move(req))
+                .then(
+                  [rep = std::move(rep)](json_status_return_type res) mutable {
+                      rep->set_status(res.status);
+                      if (res._body_writer) {
+                          rep->write_body("json", std::move(res._body_writer));
+                      } else {
+                          rep->_content += res._res;
+                      }
+                      return ss::make_ready_future<
+                        std::unique_ptr<ss::http::reply>>(std::move(rep));
```

Comment:
I couldn't think of a nicer way of wrapping this: https://github.com/redpanda-data/seastar/blob/v24.2.x/include/seastar/http/function_handlers.hh#L99-L111

Reply:
Not questioning that it is necessary, but what makes these endpoints unique to the other ones?

Reply:
> Not questioning that it is necessary, but what makes these endpoints unique to the other ones?

Typically errors are handled like this:
```c++
    if(res.has_error()) {
        ss::httpd::base_exception(
          res.assume_error().message(),
          ss::http::reply::status_type::unprocessable_entity);
    }
```
Which results in:
```json
{"message": "Request body is not a JSON object", "code": 422}
```
So it inextricably links the http status_code to the code in the message.

I'll have a look at implementing another handler type; I'll probably need it to serve a file, anyway.

Reply:
Done